### DISCUSSION TITLE
X-plane support discussion and brainstorming

### DISCUSSION THREAD

**Discussion by tigert** [MobiFlight/MobiFlight-Connector#647] [General]
Lets put a stake on the ground so we have a place to refer to.

**Comment by tigert** [comment]
Some links I found that might or might not be useful, given my limited understanding of coding and architecture.

## [XPlaneConnector](https://github.com/MaxFerretti/XPlaneConnector)
 
A friend of mine (FlyingFoxFIN on MF discord) has used this with C# to make his own code connect to X-plane. He said he is happy to answer questions etc on discord, but he is not familiar with MobiFlight code and architecture. 

He said XPlaneConnector is quite similar to Simconnect. They seem to have a hardcoded list of datarefs though, maybe it is not easy to query them all from outside the sim?

## [Datareftool](https://github.com/leecbaker/datareftool)

The dataref lookup plugin is also open source, might be useful reference. It does a query of all datarefs for example.

 * MIT license

## [XPNet](https://github.com/xpnteam/xpnet)

This is a framework that lets you write X-plane plyg-ins with .NET. Maybe a simple plugin could do a dataref lookup at startup like datareftool, and pass it to MobiFlight?

 * MIT license
 * .NET Core

**Comment by tigert** [comment]
We discussed with @pizman about x-plane datarefs, and XPUIPC config file, and thanks to discord user Samhaho we realized it is possible to encode several x-plane commands into one offset by just setting different values in Mobiflight. So like if you set OnPress event with a value of 1, you get command1, value 2 gives you command2 and so on, up to 255 commands per offset.

I also looked into my X-Plane folders and under X-Plane 11\Resources\Plugins there are a few files:

## Commands.txt

This contains "events" like this:

```sim/GPS/g1000n1_fms_outer_up                       G1000 pilot FMS outer ring tune up.
sim/GPS/g1000n1_fms_outer_down                     G1000 pilot FMS outer ring tune down.
sim/GPS/g1000n1_fms_inner_up                       G1000 pilot FMS inner ring tune up.
sim/GPS/g1000n1_fms_inner_down                     G1000 pilot FMS inner ring tune down.
```
.. containing more or less all the events in the base X-Plane. 

## Datarefs.txt

This is another file that is in the same folder. Contains stuff like this:

```sim/cockpit/radios/com1_freq_hz	int	y	10Hz	The current frequency of the copm1 radio.
sim/cockpit/radios/com2_freq_hz	int	y	10Hz	The current frequency of the com2 radio.
sim/cockpit/electrical/landing_lights_on	int	y	bool	Landing Light
sim/cockpit/electrical/nav_lights_on	int	y	bool	Navigation Light
```

Also, the popular Zibo 737-800 mod contains its own custom datarefs, that have even more useful stuff for home cockpits, things that are not implemented in the base simulator.

## Thoughts

So basically there are input and output dataref / command lists. We started to think, what if HubHop was used to generate a full XPUIPCOffsets.cfg for MobiFlight, and we would also generate a set of X-Plane presets? So that you could generate similar list of presets like for MSFS2020, but for X-plane and just have it pipe through XPUIPC for now?

Later, when we have a better support for X-Plane, we could just change the implementation "under the hood" and the user would see no difference. 

But we could collect these also in HubHop for all the add-on planes as well?
